c/src/core/encoder.c

```diff
@@ -336,8 +336,10 @@ static int pni_encoder_exit(void *ctx, pn_data_t *data, pni_node_t *node)
   pn_encoder_t *encoder = (pn_encoder_t *) ctx;
   char *pos;
 
-  // Special case 0 length list
-  if (node->atom.type==PN_LIST && node->children-encoder->null_count==0) {
+  // Special case 0 length list, but not as first element in an array
+  pni_node_t *parent = pn_data_node(data, node->parent);
+  if (node->atom.type==PN_LIST && node->children-encoder->null_count==0 &&
+      !(pn_is_in_array(data, parent, node) && pn_is_first_in_array(data, parent, node))) {
```

Comment:
I think this should test for *any* element of an array not just the first. This is because the type opcode is the array opcode for every list element and so shouldn't be changed for a 0 element list in any place in the array.

Reply:
Agreed. You are correct, this needs to happen for ALL list array members that are empty. I have checked in an update and tested it with empty lists as several members of the array including first, and all now work correctly.

Reply:
The type opcode for array elements is only once, and is determined by the type of the first element (hence the original check). However, the size/count fields for each element (even empty lists) must also be present as the data portion of each array element, and it is these that are omitted for all empty lists if this patch update is not made. Thanks for spotting that!